File: ics-watch.py
#!/usr/bin/env python3
# SPDX-License-Identifier: GPL-2.0-or-later
#
# Copyright (C) 2025 Raspberry Pi Ltd.
#
# “ICS” = Internet Connection Sharing (host-side NAT/DHCP).
#
import time, subprocess, sys, socket, struct
import os, gi
gi.require_version("NM", "1.0")
from gi.repository import GLib, NM


ICS_DEBUG = os.environ.get("ICS_DEBUG", "0") != "0"
#IFACE = os.environ.get("USB_GADGET_IFACE", "usb0")
IFACE = "usb0"
#CLIENT_ID = os.environ.get("CLIENT_NAME", "USB Gadget (client)")
#SHARED_ID = os.environ.get("SHARED_NAME", "USB Gadget (shared)")
CLIENT_ID = "USB Gadget (client)"
SHARED_ID = "USB Gadget (shared)"

# win/mac/linux defaults
ICS_GWS = ["192.168.137.1", "192.168.2.1", "10.42.0.1"]

# Tunables
LOOP_MS = 4000          # periodic check in ms
FALLBACK_DELAY = 5      # in s
MINDWELL = 1            # in s
GW_STABLE = 3           # in s
GW_UNREACH_GRACE = 1    # in s
PROBE_EVERY = 12        # in s
PROBE_TIMEOUT = 4       # in s

last_switch = 0
last_link_up = 0
last_probe = 0
last_gw_ok = 0
client: NM.Client | None = None


def now() -> int: return int(time.time())


def log(*a, force: bool = False) -> None:
    if ICS_DEBUG or force:
        print(f"[ics-watch | {time.strftime('%Y-%m-%d %H:%M:%S')}]", *a, flush=True)


def device() -> NM.Device | None:
    d = client.get_device_by_iface(IFACE)
    if d and d.get_device_type() == NM.DeviceType.ETHERNET:
        return d
    return None


def active_con_name(dev: NM.Device) -> str:
    assert dev is not None
    ac: NM.ActiveConnection | None = dev.get_active_connection()
    if not ac: return ""
    s: NM.RemoteConnection = ac.get_connection()
    return s and s.get_id() or ""


def conn_by_id(name: str) -> NM.Connection | None:
    for c in client.get_connections():
        if c.get_id() == name:
            return c
    return None


def up(name: str) -> None:
    """Activate the named connection (idempotent)."""
    c = conn_by_id(name)
    if not c:
        log(f"up(): connection '{name}' not found")
        return
    dev = device()
    if not dev:
        log("up(): device not found")
        return

    def _cb(cli, result, user_data):
        try:
            ac = cli.activate_connection_finish(result)
            if ac:
                log(f"Activated '{name}'")
        except Exception as e:
            log(f"Activate '{name}' failed: {e}")

    log(f"Activating '{name}'")
    client.activate_connection_async(c, dev, None, None, _cb, None)


def down_and_wait(name: str, timeout_ms: int = 5000) -> bool:
    """
    Deactivate the *active* connection with id `name` on IFACE.
    Blocks (nested GLib loop) until it completes or times out.
    Returns True if the connection is no longer active after completion.
    """
    dev = device()
    if not dev:
        log("down(): no device")
        return False

    ac = dev.get_active_connection()
    if not ac:
        log("down(): no active connection on device")
        return True
    if ac.get_id() != name:
        log(f"down(): '{name}' not active (active is '{ac.get_id()}'), skipping")
        return True

    try:
        client.deactivate_connection(ac, None)  # DEPRECATED
        log(f"Disconnected from '{name}'")
        return True
    except Exception as e:
        log(f"Disconnect ignored: {e}")
        return False


def _ip4_to_str(val: int | tuple[bytes, bytearray]) -> str:
    # NM’s GI returns an int (host-order) for IPv4 addresses
    if isinstance(val, int):
        return socket.inet_ntoa(struct.pack('!I', val))
    if isinstance(val, (bytes, bytearray)) and len(val) == 4:
        return socket.inet_ntoa(val)
    # already a string (fallback)
    return str(val)


def ip4_config(dev: NM.Device) -> tuple[list[str], str | None]:
    cfg = dev.get_ip4_config()
    if not cfg:
        return [], None
    addrs = []
    addrobjs = cfg.get_addresses() or []
    for a in addrobjs:
        try:
            addrs.append(_ip4_to_str(a.get_address()))
        except Exception:
            pass
    gw = cfg.get_gateway() or None
    return addrs, gw

def has_non_apipa(addrs: list[str]) -> bool:
    for a in addrs:
        if not a.startswith("169.254."):
            return True
    return False

def arping(gw: str) -> bool:
    try:
        # prefer arping; fallback to ping
        subprocess.run(["arping","-q","-c","1","-w","1","-I",IFACE,gw],
                       check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        return True
    except Exception:
        try:
            subprocess.run(["ping","-c","1","-W","1","-I",IFACE,gw],
                           check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
            return True
        except Exception:
            return False

def any_ics_gateway_reachable() -> bool:
    return any(arping(gw) for gw in ICS_GWS)

def maybe_switch(target: str) -> None:
    """Switch profiles with anti-flap and a tiny disconnect->activate handoff."""
    global last_switch
    dev = device()
    if not dev:
        log("maybe_switch(): device not found")
        return
    active = active_con_name(dev)
    if active == target:
        log(f"maybe_switch(): already on {target}")
        return
    if now() - last_switch < MINDWELL:
        log(f"maybe_switch(): switched too recently ({now()-last_switch}s ago)")
        return

    other = CLIENT_ID if target == SHARED_ID else SHARED_ID
    log(f"Switching to {target}")
    # drop the current attachment (if it's the other profile)
    if down_and_wait(other):
        last_switch = now()
        up(target)
        if target == CLIENT_ID:
            log("ICS Gateway detected; switched to DHCP client mode", force=True)
        else:
            log("No ICS Gateway detected; switched to shared mode", force=True)
        return
    log("maybe_switch(): failed to drop other profile; not switching")


def client_probe() -> None:
    """Briefly ensure client is up; if no gw appears within timeout, revert to shared."""
    global last_probe, last_switch, last_gw_ok
    if now() - last_probe < PROBE_EVERY:
        return
    last_probe = now()
    log("Client probe: trying DHCP in CLIENT")
    up(CLIENT_ID)
    t0 = now()
    while now() - t0 < PROBE_TIMEOUT:
        dev = device()
        addrs, gw = ip4_config(dev) if dev else ([], None)
        if gw and arping(gw):
            log("Client probe succeeded; staying CLIENT")
            last_switch = now(); last_gw_ok = now()
            return
        time.sleep(1)
    log("Client probe failed; reverting to SHARED")
    up(SHARED_ID); last_switch = now()


def carrier_up(dev: NM.Device) -> bool:
    try:
        return bool(dev.get_carrier())  # NMDeviceEthernet
    except Exception:
        # Fallback: consider “activated” as effectively up for our use-case
        return dev.get_state() == NM.DeviceState.ACTIVATED


def periodic_check() -> bool:
    global last_link_up, last_gw_ok
    log("Periodic check")

    dev = device()
    if not dev:
        log(f"No device {IFACE}")
        return True
    # Track link
    if carrier_up(dev):
        log("Link is up")
        if last_link_up == 0:
            last_link_up = now()
    else:
        log("Link is down")
        return True

    name = active_con_name(dev)
    addrs, gw = ip4_config(dev)
    # CLIENT mode
    if name == CLIENT_ID:
        log(f"CLIENT: addrs={addrs} gw={gw}")
        if gw and arping(gw):
            log("CLIENT: gateway reachable")
            last_gw_ok = now()
            return True
        # Gw unreachable
        log("CLIENT: gateway unreachable")
        since_ok = now() - (last_gw_ok or last_link_up)
        if since_ok >= GW_UNREACH_GRACE:
            #log(f"CLIENT: gateway lost for {since_ok}s; trying renew before fallback")
            #up(CLIENT_ID)
            #t0 = now()
            #while now() - t0 < PROBE_TIMEOUT:
            #    dev2 = device()
            #    _, gw2 = ip4_config(dev2)
            #    if gw2 and arping(gw2):
            #        log("CLIENT: renew succeeded; staying CLIENT")
            #        last_gw_ok = now()
            #        return True
            #    time.sleep(1)
            log("CLIENT: no gateway after renew; switching to SHARED")
            maybe_switch(SHARED_ID)
        elif (now()-last_link_up) >= FALLBACK_DELAY and not has_non_apipa(addrs):
            log("CLIENT: APIPA only; switching to SHARED")
            maybe_switch(SHARED_ID)
        else:
            log("CLIENT: waiting")
        return True

    # SHARED mode
    if name == SHARED_ID:
        log(f"SHARED: addrs={addrs} gw={gw}")
        if any_ics_gateway_reachable():
            log("SHARED: ICS gw detected; switching to CLIENT")
            maybe_switch(CLIENT_ID)
        else:
            log("SHARED: no ICS gw detected; staying SHARED")
        return True

    # No profile active yet: prefer CLIENT
    up(CLIENT_ID)
    last_link_up = now()
    return True

#def on_dev_state_changed(dev, old, new, reason):
#    # When NM changes state (link up/down, activation), re-evaluate soon
#    GLib.timeout_add_seconds(1, periodic_check)

if __name__ == "__main__":
    loop = GLib.MainLoop()
    client = NM.Client.new(None)
    #d = device()
    #if d:
    #    d.connect("state-changed", on_dev_state_changed)
    log(f"Starting ICS watcher on {IFACE}")
    GLib.timeout_add(LOOP_MS, periodic_check)

    import signal
    def _sigint(_s, _f):
        log("SIGINT received, exiting")
        loop.quit()
        exit(0)
    signal.signal(signal.SIGINT, _sigint)

    loop.run()

File: ics-watch.c
// SPDX-License-Identifier: Apache-2.0
// ICS watcher: toggle between "USB Gadget (client)" and "USB Gadget (shared)"
// using NetworkManager (libnm) + GLib
// ICS = "Internet Connection Sharing"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>

#include <glib.h>
#include <glib/gprintf.h>
#include <gio/gio.h>
#include <NetworkManager.h>

static const char *ENV_DEBUG = "ICS_DEBUG";
static const char *ENV_IFACE = "USB_GADGET_IFACE";
static const char *IFACE = "usb0";
static const char *CLIENT_ID = "USB Gadget (client)";
static const char *SHARED_ID = "USB Gadget (shared)";

// Win/mac/linux ICS gateway defaults
static const char *ICS_GWS[] = { "192.168.137.1", "192.168.2.1", "10.42.0.1" };
static const size_t ICS_GWS_N = 3;

// Tunables (seconds/ms)
static const guint  LOOP_MS             = 4000; // main loop tick
static const gint64 CLIENT_PROBE_WINDOW = 15;   // seconds we allow DHCP to succeed after switching to CLIENT
static const gint64 BACKOFF_AFTER_FAIL  = 15;   // seconds to wait in SHARED after a failed CLIENT try
static const gint64 MINDWELL            = 2;    // anti-flap
static const gint64 GW_LOSS_GRACE       = 10;   // how long to tolerate a dead gateway in CLIENT

static gboolean     g_debug     = FALSE;
static GMainLoop    *g_loop     = NULL;
static NMClient     *g_client   = NULL;

// State
static gint64 last_switch       = 0;
static gint64 client_deadline   = 0;  // 0 = not probing; otherwise time when CLIENT must have lease/gw
static gint64 backoff_until     = 0;  // 0 = no backoff; otherwise do not try CLIENT before this time
static gint64 last_gw_ok        = 0;  // last time we saw a working gateway in CLIENT mode

// --- helpers -----------------------------------------------------------------

static inline gint64 now_s(void) {
    return g_get_real_time() / G_USEC_PER_SEC;
}

static void log_msg(gboolean force, const char *fmt, ...) {
    if (!g_debug && !force) return;
    va_list ap;
    va_start(ap, fmt);
    GDateTime *dt = g_date_time_new_now_local();
    gchar *ts = g_date_time_format(dt, "%F %T");
    g_print("[ics-watch | %s] ", ts);
    g_vprintf(fmt, ap);
    g_print("\n");
    g_free(ts);
    g_date_time_unref(dt);
    va_end(ap);
}

static NMDevice *get_device(void) {
    NMDevice *d = nm_client_get_device_by_iface(g_client, IFACE);
    if (!d) {
        return NULL;
    }
    if (nm_device_get_device_type(d) == NM_DEVICE_TYPE_ETHERNET) {
        return d;
    }

    return NULL;
}

static const char *get_active_con_id(NMDevice *dev) {
    NMActiveConnection *ac = nm_device_get_active_connection(dev);
    if (!ac) {
        return "";
    }
    const char *id = nm_active_connection_get_id(ac);
    return id ? id : "";
}

static NMConnection *conn_by_id(const char *id) {
    const GPtrArray *conns = nm_client_get_connections(g_client);
    for (guint i = 0; i < conns->len; i++) {
        NMConnection *c = g_ptr_array_index((GPtrArray *)conns, i);
        const char *cid = nm_connection_get_id(c);
        if (cid && g_strcmp0(cid, id) == 0) return c;
    }
    return NULL;
}

static gboolean is_active_id(const char *want_id) {
    NMDevice *dev = get_device();
    if (!dev)
        return FALSE;
    const char *id = get_active_con_id(dev);
    return id && g_strcmp0(id, want_id) == 0;
}

static gboolean is_transitioning(NMDevice *dev) {
    NMDeviceState s = nm_device_get_state(dev);
    return (s > NM_DEVICE_STATE_DISCONNECTED && s < NM_DEVICE_STATE_ACTIVATED);
}

typedef struct { char *name; } ActivateCtx;

static void on_activate_cb(GObject *src, GAsyncResult *res, gpointer user_data) {
    ActivateCtx *ctx = user_data;
    GError *err = NULL;
    NMActiveConnection *ac =
        nm_client_activate_connection_finish(NM_CLIENT(src), res, &err);
    if (!ac) {
        log_msg(FALSE, "Activate '%s' failed: %s", ctx->name, err ? err->message : "unknown");
        g_clear_error(&err);
    } else {
        log_msg(FALSE, "Activated '%s'", ctx->name);
    }
    g_free(ctx->name);
    g_free(ctx);
}

static gboolean up(const char *name) {
    NMDevice *dev = get_device();
    if (!dev) {
        log_msg(FALSE, "up(): no device");
        return FALSE;
    }

    if (is_active_id(name)) {
        log_msg(FALSE, "'%s' already active", name);
        return TRUE;
    } else if (is_transitioning(dev)) {
        log_msg(FALSE, "Activation already in progress (state=%d); skip up('%s')",
                nm_device_get_state(dev), name);
        return TRUE; // treat as fine
    }
    
    NMConnection *c = conn_by_id(name);
    if (!c) {
        log_msg(FALSE, "up(): connection '%s' not found", name);
        return FALSE;
    }

    ActivateCtx *ctx = g_new0(ActivateCtx, 1);
    ctx->name = g_strdup(name);

    nm_client_activate_connection_async(
        g_client, c, dev, NULL, NULL, on_activate_cb, ctx);
    
    log_msg(FALSE, "Activating '%s'...", name);
    return TRUE;
}

typedef struct { GMainLoop *loop; gboolean ok; } DeactCtx;

static void on_deact_cb(GObject *src, GAsyncResult *res, gpointer user_data) {
    DeactCtx *ctx = user_data;
    GError *err = NULL;
    ctx->ok = nm_client_deactivate_connection_finish(NM_CLIENT(src), res, &err);
    if (!ctx->ok) {
        log_msg(FALSE, "Deactivate failed: %s", err ? err->message : "unknown");
        g_clear_error(&err);
    } else {
        log_msg(FALSE, "Deactivated connection");
    }
    if (ctx->loop)
        g_main_loop_quit(ctx->loop);
}

static gboolean quit_loop_cb(gpointer data) {
    g_main_loop_quit((GMainLoop *)data);
    return G_SOURCE_REMOVE;
}

static gboolean down_and_wait(const char *name, guint timeout_ms) {
    NMDevice *dev = get_device();
    if (!dev) {
        log_msg(FALSE, "down(): no device");
        return FALSE;
    }

    NMActiveConnection *ac = nm_device_get_active_connection(dev);
    if (!ac) {
        log_msg(FALSE, "down(): no active connection on device");
        return TRUE;
    }

    const char *active_id = nm_active_connection_get_id(ac);
    if (!active_id || g_strcmp0(active_id, name) != 0) {
        log_msg(FALSE, "down(): '%s' not active (active is '%s'), skipping",
             name, active_id ? active_id : "");
        return TRUE;
    }

    DeactCtx ctx = {0};
    ctx.loop = g_main_loop_new(NULL, FALSE);

    nm_client_deactivate_connection_async(g_client, ac, NULL, on_deact_cb, &ctx);

    guint to = g_timeout_add(timeout_ms, quit_loop_cb, ctx.loop);
    g_main_loop_run(ctx.loop);
    g_source_remove(to);
    g_main_loop_unref(ctx.loop);

    return ctx.ok;
}

static gboolean carrier_up(NMDevice *dev) {
    if (NM_IS_DEVICE_ETHERNET(dev)) {
        NMDeviceEthernet *eth = NM_DEVICE_ETHERNET(dev);
        return nm_device_ethernet_get_carrier(eth);
    }
    return nm_device_get_state(dev) == NM_DEVICE_STATE_ACTIVATED;
}

static void ip4_config(NMDevice *dev, GPtrArray **addrs_out, const char **gw_out) {
    *gw_out = NULL;
    *addrs_out = g_ptr_array_new_with_free_func(g_free);

    NMIPConfig *cfg = nm_device_get_ip4_config(dev);
    if (!cfg)
        return;

    const GPtrArray *addrs = nm_ip_config_get_addresses(cfg);
    if (addrs) {
        for (guint i = 0; i < addrs->len; i++) {
            NMIPAddress *ipa = g_ptr_array_index((GPtrArray *)addrs, i);
            const char *s = nm_ip_address_get_address(ipa);
            if (s)
                g_ptr_array_add(*addrs_out, g_strdup(s));
        }
    }
    const char *gw = nm_ip_config_get_gateway(cfg);
    if (gw) *gw_out = gw;
}

static gboolean has_non_apipa(GPtrArray *addrs) {
    for (guint i = 0; i < addrs->len; i++) {
        const char *s = addrs->pdata[i];
        if (g_str_has_prefix(s, "169.254."))
            continue;
        return TRUE;
    }
    return FALSE;
}

// check if is gateway reachable via ARP
static gboolean arping(const char *gw) {
    gchar *arp = g_find_program_in_path("arping");
    if (!arp) {
        log_msg(FALSE, "arping not found in PATH");
        return FALSE;
    }

    gchar *argv1[] = { arp, "-q", "-c", "1", "-w", "1",
                        "-I", (gchar*)IFACE, (gchar*)gw, NULL };
    GError *err = NULL;
    gint status = 0;
    gboolean ok = FALSE;

    if (g_spawn_sync(NULL, argv1, NULL,
                     G_SPAWN_STDOUT_TO_DEV_NULL |
                     G_SPAWN_STDERR_TO_DEV_NULL,
                     NULL, NULL, NULL, NULL, &status, &err)) {
        ok = g_spawn_check_wait_status(status, NULL);
    } else {
        log_msg(FALSE, "arping spawn failed: %s", err ? err->message : "unknown");
        g_clear_error(&err);
    }

    g_free(arp);
    return ok;
}

static gboolean ics_gw_is_reachable(void) {
    for (size_t i = 0; i < ICS_GWS_N; i++)
        if (arping(ICS_GWS[i]))
            return TRUE;
    return FALSE;
}

static void maybe_switch(const char *target) {
    NMDevice *dev = get_device();
    if (!dev) {
        log_msg(FALSE, "maybe_switch(): device not found");
        return;
    }

    const char *active = get_active_con_id(dev);
    if (g_strcmp0(active, target) == 0) {
        log_msg(FALSE, "maybe_switch(): already on %s", target);
        return;
    }
    gint64 t = now_s();
    if (t - last_switch < MINDWELL) {
        log_msg(FALSE, "maybe_switch(): switch too recent");
        return;
    }

    const char *other = (g_strcmp0(target, SHARED_ID) == 0) ? CLIENT_ID : SHARED_ID;
    log_msg(FALSE, "Switching to %s", target);

    if (down_and_wait(other, 5000)) {
        last_switch = now_s();
        if (up(target)) {
            if (g_strcmp0(target, CLIENT_ID) == 0) {
                client_deadline = now_s() + CLIENT_PROBE_WINDOW;
                last_gw_ok = 0; // <- reset; we haven't proved the GW yet
                log_msg(TRUE, "Trying DHCP client mode (deadline in %lds)", (long)CLIENT_PROBE_WINDOW);
            } else {
                log_msg(TRUE, "Switched to shared mode");
            }
        }
    } else {
        log_msg(FALSE, "failed to drop other profile; not switching");
    }
}

// --- main loop -------------------------------------------------------------

static gboolean periodic_check(gpointer user_data) {
    (void)user_data;

    NMDevice *dev = get_device();
    if (!dev) {
        log_msg(FALSE, "No device %s", IFACE);
        return G_SOURCE_CONTINUE;
    } else if (!carrier_up(dev)) {
        log_msg(FALSE, "Link down");
        return G_SOURCE_CONTINUE;
    }

    const char *name = get_active_con_id(dev);
    GPtrArray *addrs = NULL;
    const char *gw = NULL;
    ip4_config(dev, &addrs, &gw);

    if (g_strcmp0(name, CLIENT_ID) == 0) {
        // CLIENT mode
        const gboolean ok_ip = has_non_apipa(addrs);
        const gint64   now   = now_s();

        // If we don't even have a non-APIPA or a default GW yet, keep probing until deadline expires.
        if (!ok_ip || !gw) {
            if (client_deadline == 0)
                client_deadline = now + CLIENT_PROBE_WINDOW;
            const gint64 left = client_deadline - now;
            gchar *joined = g_strjoinv(",", (gchar **)addrs->pdata);
            if (left > 0) {
                log_msg(FALSE, "CLIENT waiting: addrs=%s gw=%s (deadline %lds)",
                        joined ? joined : "(…)", gw ? gw : "(none)", (long)left);
            } else {
                log_msg(FALSE, "CLIENT failed (APIPA/no GW); back to SHARED, backoff %lds",
                        (long)BACKOFF_AFTER_FAIL);
                backoff_until = now + BACKOFF_AFTER_FAIL;
                maybe_switch(SHARED_ID);
            }

            g_free(joined);
            g_ptr_array_free(addrs, TRUE);
            return G_SOURCE_CONTINUE;
        }

        // We have non-APIPA and a GW configured — validate that the GW actually answers ARP.
        if (arping(gw)) {
            last_gw_ok = now;
            gchar *joined = g_strjoinv(",", (gchar **)addrs->pdata);
            log_msg(FALSE, "CLIENT OK: addrs=%s gw=%s", joined ? joined : "(…)", gw);
            g_free(joined);
            g_ptr_array_free(addrs, TRUE);
            return G_SOURCE_CONTINUE;
        }

        // GW not responding; hold CLIENT for a short grace, then fall back to SHARED.
        if (last_gw_ok == 0)
            last_gw_ok = now; // start the grace window the first time it fails
        const gint64 since_ok = now - last_gw_ok;
        const gint64 left = (since_ok < GW_LOSS_GRACE) ? (GW_LOSS_GRACE - since_ok) : 0;

        if (left > 0) {
            gchar *joined = g_strjoinv(",", (gchar **)addrs->pdata);
            log_msg(FALSE, "CLIENT: GW %s not responding; grace %lds left (addrs=%s)",
                    gw, (long)left, joined ? joined : "(…)");
            g_free(joined);
        } else {
            log_msg(FALSE, "CLIENT: GW %s lost for >=%lds; back to SHARED (backoff %lds)",
                    gw, (long)GW_LOSS_GRACE, (long)BACKOFF_AFTER_FAIL);
            backoff_until = now + BACKOFF_AFTER_FAIL;
            maybe_switch(SHARED_ID);
        }

        g_ptr_array_free(addrs, TRUE);
        return G_SOURCE_CONTINUE;
    }
    
    if (g_strcmp0(name, SHARED_ID) == 0) {
        // SHARED: stay put unless backoff expired AND ICS GW looks present -> try CLIENT once
        gchar *joined = g_strjoinv(",", (gchar **)addrs->pdata);
        log_msg(FALSE, "SHARED: addrs=%s", joined ? joined : "(…)"); g_free(joined);

        if (now_s() >= backoff_until && ics_gw_is_reachable()) {
            log_msg(FALSE, "SHARED: ICS GW responded; trying CLIENT");
            maybe_switch(CLIENT_ID);
        } else if (now_s() < backoff_until) {
            log_msg(FALSE, "SHARED: backoff active (%lds left)", (long)(backoff_until - now_s()));
        } else {
            log_msg(FALSE, "SHARED: no ICS GW; staying");
        }

        g_ptr_array_free(addrs, TRUE);
        return G_SOURCE_CONTINUE;
    }

    // No profile yet: start by trying CLIENT once (then logic above handles fallback/backoff)
    if (!is_transitioning(dev)) {
        up(CLIENT_ID);
        client_deadline = now_s() + CLIENT_PROBE_WINDOW;
        last_switch = now_s();
    } else {
        log_msg(FALSE, "Activation in progress; not reissuing up()");
    }

    g_ptr_array_free(addrs, TRUE);
    return G_SOURCE_CONTINUE;
}

// --- boilerplate -----------------------------------------------------------

static void on_sigint(int signum) {
    (void)signum;
    log_msg(TRUE, "SIGINT received, exiting");
    if (g_loop)
        g_main_loop_quit(g_loop);
}

int main(int argc, char **argv) {
    (void)argc; (void)argv;

    const char *dbg = g_getenv(ENV_DEBUG);
    g_debug = (dbg && strcmp(dbg, "0") != 0);

    const char *iface_env = g_getenv(ENV_IFACE);
    if (iface_env && *iface_env)
        IFACE = iface_env;

    GError *err = NULL;
    g_client = nm_client_new(NULL, &err);
    if (!g_client) {
        g_printerr("Failed to create NMClient: %s\n", err ? err->message : "unknown error");
        g_clear_error(&err);
        return 1;
    }

    log_msg(TRUE, "Starting ICS watcher (debug=%d) on %s", g_debug, IFACE);

    g_loop = g_main_loop_new(NULL, FALSE);
    g_timeout_add(LOOP_MS, periodic_check, NULL);

    // SIGINT
    struct sigaction sa;
    memset(&sa, 0, sizeof(sa));
    sa.sa_handler = on_sigint;
    sigaction(SIGINT, &sa, NULL);

    g_main_loop_run(g_loop);

    g_main_loop_unref(g_loop);
    g_clear_object(&g_client);
    return 0;
}
